services/traces/downloader.go

```diff
@@ -0,0 +1,91 @@
+package traces
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"time"
+
+	"cloud.google.com/go/storage"
+	lru "github.com/hashicorp/golang-lru/v2"
+	"github.com/onflow/go-ethereum/common"
+	"github.com/rs/zerolog"
+)
+
+const downloadTimeout = 60 * time.Second
+
+type Downloader interface {
+	// Download traces or returning an error with the failure
+	Download(id common.Hash) (json.RawMessage, error)
+}
+
+var _ Downloader = &GCPDownloader{}
+
+type GCPDownloader struct {
+	client     *storage.Client
+	logger     zerolog.Logger
+	cache      *lru.TwoQueueCache[string, []byte]
+	bucketName string
+}
+
+func NewGCPDownloader(bucketName string, logger zerolog.Logger) (*GCPDownloader, error) {
+	if bucketName == "" {
+		return nil, fmt.Errorf("must provide bucket name")
+	}
+
+	ctx := context.Background()
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("storage.NewClient: %w", err)
+	}
+
+	// try accessing buckets to validate settings
+	_, err = client.Bucket(bucketName).Attrs(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error accessing bucket: %s, make sure bucket exists: %w", bucketName, err)
+	}
+
+	valueCache, err := lru.New2Q[string, []byte](128)
```

Comment:
caching in the downloader seems strange. I feel like this belongs in the API or ingestion engine